pkg/render/manager_route_config.go

```diff
@@ -0,0 +1,396 @@
+		volumeMounts: builder.volumeMounts,
+		volumes:      builder.volumes,
+		annotation:   rmeta.AnnotationHash(builder.hashableContent),
+	}, nil
+}
+
+func (builder *voltronRouteConfigBuilder) mountConfigMapReference(name, key string) (string, error) {
+	defaultMode := int32(420)
+
+	configMap := builder.configMaps[name]
+	if configMap == nil {
+		return "", fmt.Errorf("the contents for ConfigMap '%s' wasn't provide, and is needed to generate annotations", name)
+	}
+
+	if _, ok := builder.mountedConfigMaps[name]; !ok {
+		// Prefix the volume name to avoid collisions with secrets that have the same name.
+		volumeName := fmt.Sprintf("cm-%s", name)
+
+		// Use a different folder for secrets and configmaps to avoid mounting collisions.
+		mountLocation := fmt.Sprintf("%s/%s", configMapFolder, name)
+		volumeMount := corev1.VolumeMount{Name: volumeName, MountPath: mountLocation, ReadOnly: true}
+
+		volume := corev1.Volume{
+			Name: volumeName,
+			VolumeSource: corev1.VolumeSource{
+				ConfigMap: &corev1.ConfigMapVolumeSource{
+					LocalObjectReference: corev1.LocalObjectReference{Name: name},
+					DefaultMode:          &defaultMode,
+				},
+			},
+		}
+
+		builder.volumeMounts = append(builder.volumeMounts, volumeMount)
+		builder.volumes = append(builder.volumes, volume)
+
+		// Use mapToSortedArray to ensure the ordering is the same and the annotation value isn't changed because of
+		// map ordering differences.
+		builder.hashableContent = append(builder.hashableContent, mapToSortedArray(configMap.Data))
+		builder.mountedConfigMaps[name] = struct{}{}
+	}
+
+	return fmt.Sprintf("%s/%s/%s", configMapFolder, name, key), nil
+}
+
+func (builder *voltronRouteConfigBuilder) mountSecretReference(name, key string) (string, error) {
+	defaultMode := int32(420)
+
+	secret := builder.secrets[name]
+
+	if secret == nil {
+		return "", fmt.Errorf("the contents for Secret '%s' wasn't provide, and is needed to generate annotations", name)
+	}
+
+	if _, ok := builder.mountedSecrets[name]; !ok {
+		// Prefix the volume name to avoid collisions with secrets that have the same name.
+		volumeName := fmt.Sprintf("scrt-%s", name)
+
+		// Use a different folder for secrets and configmaps to avoid mounting collisions.
+		mountLocation := fmt.Sprintf("%s/%s", secretsFolder, name)
+		volumeMount := corev1.VolumeMount{Name: volumeName, MountPath: mountLocation, ReadOnly: true}
+
+		volume := corev1.Volume{
+			Name: volumeName,
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{
+					SecretName:  name,
+					DefaultMode: &defaultMode,
+				},
+			},
+		}
+
+		builder.volumeMounts = append(builder.volumeMounts, volumeMount)
+		builder.volumes = append(builder.volumes, volume)
+
+		// Use mapToSortedArray to ensure the ordering is the same and the annotation value isn't changed because of
+		// map ordering differences.
+		builder.hashableContent = append(builder.hashableContent, mapToSortedArray(secret.Data))
```

Comment:
Wonder if we could avoid this "incremental" build of the hashable content (which makes it a bit harder to follow and debug if needed) and instead build this all at once at the end by just looping over the secrets / configmaps. Would also remove the need for the extra struct field.

In other similar places within the operator, we generally do an annotation per-dependency so that it's easier to tell what changed. If we hash all of these into a single annotation, it becomes harder to monitor and see which dependent object is triggering the update.

Reply:
This is definitely a bit more complex than needed, you're right that we don't need this incremental build.

I was thinking that we'd want to avoid correlating the number of annotations to something that's unbounded and configurable outside of the operator (we could have any number of tls routes configured). Practically speaking though, I doubt we'd have hundreds of tls routes configured. So if you'd rather err on the side of ease of debugging then I'm fine with that (just comment again saying so please).